lib/invitations/inviteUserByEmail.ts

```diff
@@ -0,0 +1,208 @@
+import { FormRecord, UserAbility } from "@lib/types";
+import { getUser } from "@lib/users";
+import {
+  InvalidDomainError,
+  MismatchedEmailDomainError,
+  TemplateNotFoundError,
+  UserAlreadyHasAccessError,
+  UserNotFoundError,
+} from "./exceptions";
+import { getTemplateWithAssociatedUsers } from "@lib/templates";
+import { prisma } from "@lib/integration/prismaConnector";
+import { sendEmail } from "@lib/integration/notifyConnector";
+import { inviteToCollaborate } from "@lib/invitations/emailTemplates/inviteToCollaborate";
+import { inviteToForms } from "@lib/invitations/emailTemplates/inviteToForms";
+import { getOrigin } from "@lib/origin";
+import { logMessage } from "@lib/logger";
+import { Invitation } from "@prisma/client";
+import { logEvent } from "@lib/auditLogs";
+import { isValidGovEmail } from "@lib/validation/validation";
+
+/**
+ * Invite someone to the form by email
+ *
+ * @param ability
+ * @param email
+ * @param formId
+ */
+export const inviteUserByEmail = async (
+  ability: UserAbility,
+  email: string,
+  formId: string,
+  message: string
+) => {
+  let invitation;
+
+  const sender = await getUser(ability, ability.userID);
+
+  if (!sender) {
+    throw new UserNotFoundError();
+  }
+
+  const template = await getTemplateWithAssociatedUsers(ability, formId);
+
+  if (!template) {
+    throw new TemplateNotFoundError();
+  }
+
+  // check if user is already associated with the form
+  if (template.users.some((user) => user.email === email)) {
+    throw new UserAlreadyHasAccessError();
+  }
+
+  if (!isValidGovEmail(email)) {
+    throw new InvalidDomainError();
+  }
+
+  // Check if the email domain matches the sender's email domain
+  const senderDomain = sender.email.split("@")[1];
+  const recipientDomain = email.split("@")[1];
+
+  if (senderDomain !== recipientDomain) {
+    throw new MismatchedEmailDomainError();
+  }
+
+  // check if user is already invited to the form
+  const previousInvitation = await _retrieveFormInvitationByEmail(email, formId);
+  if (previousInvitation) {
+    invitation = previousInvitation;
+    // if invitation is expired, delete and recreate
+    if (previousInvitation.expires < new Date()) {
```

Comment:
What happens to invitations that are never accepted or denied? Do the entries stay in the DB until the form is deleted?

Reply:
Good question. As it is, it would be up to the owner(s) on the form to cleanup old/expired invitations.

Open to suggestions- should we auto-clean based on some interval? Should we try to draw attention to particularly old invitations to encourage cleanup?

Reply:
I would think a cron job to clean after x amount of time would make sense.

Reply:
Do we have cron?

Reply:
Unfortunatley we don't have access to 'cron' which is why we tend to setup so many Lambda functions.  Definitely not urgent and I guess it really depends on how users behave in the app. For example do Forms actually get deleted or are the mostly abandoned? We could track this as out of scope but a follow up  with an issue that needs more research on?

Reply:
I'll create an issue to track this, as you say- see how it's used and if it becomes a problem.